File: backend/src/main/java/com/lsoft/aguapotable/controller/FacturacionController.java
package com.lsoft.aguapotable.controller;

import com.lsoft.aguapotable.service.FacturacionService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/api/v1/facturacion")
@RequiredArgsConstructor
@Tag(name = "Facturación", description = "Generación de facturas a partir de lecturas")
@SecurityRequirement(name = "bearerAuth")
public class FacturacionController {

    private final FacturacionService facturacionService;

    @PostMapping("/generar/{lecturaId}")
    @PreAuthorize("hasAnyRole('SUPER_ADMIN','ADMIN','CAJERO')")
    @Operation(summary = "Generar factura individual para una lectura")
    public ResponseEntity<?> generarFactura(@PathVariable Long lecturaId,
                                             HttpServletRequest request) {
        Long comunidadId = (Long) request.getAttribute("comunidadId");
        return ResponseEntity.status(201).body(facturacionService.generarFactura(lecturaId, comunidadId));
    }

    @PostMapping("/generar-masivo")
    @PreAuthorize("hasAnyRole('SUPER_ADMIN','ADMIN')")
    @Operation(summary = "Generación masiva de facturas para todo el período")
    public ResponseEntity<Map<String, Object>> generarMasivo(
            @RequestParam Integer anio,
            @RequestParam Integer mes,
            HttpServletRequest request) {
        Long comunidadId = (Long) request.getAttribute("comunidadId");
        int generadas = facturacionService.generarFacturasMasivas(comunidadId, anio, mes);
        return ResponseEntity.ok(Map.of("facturasGeneradas", generadas, "anio", anio, "mes", mes));
    }
}

File: backend/src/facturacion/facturacion.controller.ts
import {
  Controller,
  Post,
  Param,
  Query,
  ParseIntPipe,
  HttpCode,
  UseGuards,
} from '@nestjs/common';
import { AuthGuard } from '@nestjs/passport';
import { ApiTags, ApiOperation, ApiBearerAuth } from '@nestjs/swagger';
import { FacturacionService } from './facturacion.service';
import { ComunidadId } from '../common/decorators';
import { RolesGuard, Roles } from '../auth/guards/roles.guard';

@ApiTags('Facturación')
@ApiBearerAuth()
@Controller('api/v1/facturacion')
@UseGuards(AuthGuard('jwt'), RolesGuard)
export class FacturacionController {
  constructor(private readonly facturacionService: FacturacionService) {}

  @Post('generar/:lecturaId')
  @Roles('SUPER_ADMIN', 'ADMIN', 'CAJERO')
  @HttpCode(201)
  @ApiOperation({ summary: 'Generar factura individual para una lectura' })
  async generarFactura(
    @Param('lecturaId', ParseIntPipe) lecturaId: number,
    @ComunidadId() comunidadId: number,
  ) {
    return this.facturacionService.generarFactura(lecturaId, comunidadId);
  }

  @Post('generar-masivo')
  @Roles('SUPER_ADMIN', 'ADMIN')
  @ApiOperation({ summary: 'Generación masiva de facturas para todo el período' })
  async generarMasivo(
    @Query('anio', ParseIntPipe) anio: number,
    @Query('mes', ParseIntPipe) mes: number,
    @ComunidadId() comunidadId: number,
  ) {
    const generadas = await this.facturacionService.generarFacturasMasivas(
      comunidadId,
      anio,
      mes,
    );
    return { facturasGeneradas: generadas, anio, mes };
  }
}
